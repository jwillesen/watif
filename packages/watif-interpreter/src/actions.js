import Immutable from 'immutable'
import { createActions } from 'redux-actions'

export const {
  setItemState,
  setCurrentRoom,
  setCurentItem,
  computeDisplayValues,
  executeVerb,

} = createActions({

/* eslint-disable indent */
SET_ITEM_STATE: (itemId, newItemState) => { // (string, [Object | Immutable.Map])
  if (!Immutable.Map.isMap(newItemState)) {
    newItemState = Immutable.fromJS(newItemState)
  }
  return { itemId, newItemState }
},

},

'SET_CURRENT_ROOM',       // (itemId: string)
'SET_CURRENT_ITEM',       // (itemId: string)
'ADD_LOG_ENTRY',          // (entry: watext)

'EXECUTE_VERB',           // (verb invokation)

'COMPUTE_DISPLAY_VALUES'  // ()

)
/* eslint-enable indent */

// verb invokation:
//  {
//    verb: string,
//    subject: itemId,
//    object: itemId,
//  }
