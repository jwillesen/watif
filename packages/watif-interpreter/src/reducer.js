import Immutable from 'immutable'

const defaultState = Immutable.fromJS({
  // the storyState can be replaced wholesale to implement undo/redo/editing
  storyState: {
    // array of watext
    log: [],

    // watext
    currentRoomDescription: null,

    // watext
    currentItemDescription: null,

    // map itemIds to the Item's state
    itemStates: {},
  },

  // map itemIds to Item instances
  items: {},
})

export default function (state, action) {
  if (!state) return defaultState

  if (action.type === 'SET_ITEM_STATE') {
    const {itemId, newState} = action.payload
    return state.mergeDeep({states: {[itemId]: newState}})
  }
}
