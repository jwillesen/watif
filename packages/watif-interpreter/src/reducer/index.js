import Immutable from 'immutable'
import itemStatesReducer from './item-states-reducer'
import {combineReducers} from 'redux-immutable'

// verb structure:
//  {
//    verb: string,
//    compound: bool,
//    connector: string,
// }

// export const DefaultState = Immutable.Record({
export const DefaultState = () => Immutable.fromJS({
  // array of watext
  log: [],
  // itemId string
  currentItemId: null,
  // itemId string
  currentRoomId: null,
  // map itemIds to the Item's state
  itemStates: {},

  // calculated values that also need to be communicated to the display
  // because it can't calculate them itself
  // watext
  currentRoomDescription: null,

  // array of verb structures
  currentRoomVerbs: [],

  // watext
  currentItemDescription: null,

  // array of verb structures
  currentItemVerbs: [],

  // the player's inventory and other containers
  //  {
  //    inventory: { itemId: itemName, ...}
  //    otherContainer: { ... }
  //  }
  player: {},
})

function noop (state, action) { return state }

const reducer = combineReducers({
  log: noop,
  currentItemId: noop,
  currentRoomId: noop,
  itemStates: itemStatesReducer,
  currentRoomDescription: noop,
  currentItemDescription: noop,
  inventory: noop,
}, DefaultState)

export default reducer
