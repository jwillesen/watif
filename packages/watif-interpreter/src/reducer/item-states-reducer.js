export default function itemStatesReducer (state, action) {
  if (action.type === 'SET_ITEM_STATE') {
    const {itemId, newItemState} = action.payload
    return state.mergeDeep({[itemId]: newItemState})
  }
  return state
}
