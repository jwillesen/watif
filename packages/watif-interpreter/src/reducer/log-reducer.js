import Immutable from 'immutable'
import {handleActions} from 'redux-actions'

const reducer = handleActions(
  {
    ADD_LOG_ENTRY: (state, action) => state.push(action.payload),
  },
  Immutable.List()
)

export default reducer
