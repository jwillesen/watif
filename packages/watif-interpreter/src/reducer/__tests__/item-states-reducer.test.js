import Immutable from 'immutable'
import rootReducer from '../index'
import reducer from '../item-states-reducer'
import {setItemState} from '../../actions'

// TODO: I'm not sure we can count on jest's snapshot when it comes to unordered Immutable.Map

const DefaultItemStates = rootReducer(undefined, {type: '~~init~~'})
  .get('itemStates')

it('merges item states', () => {
  let nextState = reducer(DefaultItemStates,
    setItemState('master-sword', {location: 'spooky-forest'}))
  expect(nextState).toMatchSnapshot()
  nextState = reducer(nextState,
    setItemState('master-sword', {power: 42}))
  expect(nextState.toJS()).toMatchSnapshot()
})

it('deep merges item states', () => {
  const currentState = Immutable.fromJS({
    'some-item': {
      firstLevel: {
        foo: '1',
        bar: '2',
        baz: '3',
      },
    },
  })
  const result = reducer(currentState, setItemState('some-item', {
    otherFirstLevel: 'superfluous',
    firstLevel: {
      foo: '42',
      bing: '4',
    },
  }))
  expect(result.toJS()).toMatchSnapshot()
})
