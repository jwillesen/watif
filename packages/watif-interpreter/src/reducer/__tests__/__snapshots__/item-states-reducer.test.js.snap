// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep merges item states 1`] = `
Object {
  "some-item": Object {
    "firstLevel": Object {
      "bar": "2",
      "baz": "3",
      "bing": "4",
      "foo": "42",
    },
    "otherFirstLevel": "superfluous",
  },
}
`;

exports[`merges item states 1`] = `
Immutable.Map {
  master-sword: Immutable.Map {
    location: "spooky-forest",
  },
}
`;

exports[`merges item states 2`] = `
Object {
  "master-sword": Object {
    "location": "spooky-forest",
    "power": 42,
  },
}
`;
